From d23118ff3bb72e5fa1265eadf09d598444471b99 Mon Sep 17 00:00:00 2001
From: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Date: Tue, 7 Sep 2021 16:28:21 +0200
Subject: [PATCH 0437/1695] drivers: scmi-msg: implement duty cycle ops with
 SCMI clocks

Some driver from non secure need to known the duty cycle value of
their clock.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 1bbf852267d2529492cce1097e25ba8e6437e4a3 drivers: scmi-msg: implement duty cycle ops with SCMI clocks]
Change-Id: I8ae7b13329ffa663be41672932b4893b7facc5b7
Signed-off-by: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
---
 core/drivers/scmi-msg/clock.c   | 37 +++++++++++++++++++++++++++++++++
 core/drivers/scmi-msg/clock.h   | 11 ++++++++++
 core/include/drivers/scmi-msg.h |  5 +++++
 3 files changed, 53 insertions(+)

diff --git a/core/drivers/scmi-msg/clock.c b/core/drivers/scmi-msg/clock.c
index dce76fa90..778eeddd5 100644
--- a/core/drivers/scmi-msg/clock.c
+++ b/core/drivers/scmi-msg/clock.c
@@ -5,6 +5,7 @@
  */
 #include <assert.h>
 #include <confine_array_index.h>
+#include <drivers/clk.h>
 #include <drivers/scmi-msg.h>
 #include <drivers/scmi.h>
 #include <string.h>
@@ -68,6 +69,13 @@ int32_t __weak plat_scmi_clock_set_state(unsigned int channel_id __unused,
 	return SCMI_NOT_SUPPORTED;
 }
 
+int32_t __weak plat_scmi_clock_get_duty_cycle(unsigned int channel_id __unused,
+					      unsigned int scmi_id __unused,
+					      struct clk_duty *duty __unused)
+{
+	return SCMI_NOT_SUPPORTED;
+}
+
 static void report_version(struct scmi_msg *msg)
 {
 	struct scmi_protocol_version_p2a return_values = {
@@ -213,6 +221,34 @@ static void scmi_clock_rate_set(struct scmi_msg *msg)
 	scmi_status_response(msg, status);
 }
 
+static void scmi_clock_duty_cycle_get(struct scmi_msg *msg)
+{
+	const struct scmi_clock_duty_cycle_get_a2p *in_args = (void *)msg->in;
+	struct scmi_clock_duty_cycle_get_p2a return_values = { };
+	struct clk_duty duty = { 1, 2};
+	unsigned int clock_id = 0;
+
+	if (msg->in_size != sizeof(*in_args)) {
+		scmi_status_response(msg, SCMI_PROTOCOL_ERROR);
+		return;
+	}
+
+	if (in_args->clock_id >= plat_scmi_clock_count(msg->channel_id)) {
+		scmi_status_response(msg, SCMI_INVALID_PARAMETERS);
+		return;
+	}
+
+	clock_id = confine_array_index(in_args->clock_id,
+				       plat_scmi_clock_count(msg->channel_id));
+
+	plat_scmi_clock_get_duty_cycle(msg->channel_id, clock_id, &duty);
+
+	return_values.num = duty.num;
+	return_values.den = duty.den;
+
+	scmi_write_response(msg, &return_values, sizeof(return_values));
+}
+
 static void scmi_clock_config_set(struct scmi_msg *msg)
 {
 	const struct scmi_clock_config_set_a2p *in_args = (void *)msg->in;
@@ -365,6 +401,7 @@ static const scmi_msg_handler_t scmi_clock_handler_table[] = {
 	[SCMI_CLOCK_RATE_SET] = scmi_clock_rate_set,
 	[SCMI_CLOCK_RATE_GET] = scmi_clock_rate_get,
 	[SCMI_CLOCK_CONFIG_SET] = scmi_clock_config_set,
+	[SCMI_CLOCK_DUTY_CYCLE_GET] = scmi_clock_duty_cycle_get,
 };
 
 static bool message_id_is_supported(unsigned int message_id)
diff --git a/core/drivers/scmi-msg/clock.h b/core/drivers/scmi-msg/clock.h
index a20678c33..ad3714c78 100644
--- a/core/drivers/scmi-msg/clock.h
+++ b/core/drivers/scmi-msg/clock.h
@@ -23,6 +23,7 @@ enum scmi_clock_command_id {
 	SCMI_CLOCK_RATE_SET = 0x005,
 	SCMI_CLOCK_RATE_GET = 0x006,
 	SCMI_CLOCK_CONFIG_SET = 0x007,
+	SCMI_CLOCK_DUTY_CYCLE_GET = 0x00B,
 };
 
 /* Protocol attributes */
@@ -58,6 +59,16 @@ struct scmi_clock_rate_get_p2a {
 	uint32_t rate[2];
 };
 
+struct scmi_clock_duty_cycle_get_a2p {
+	uint32_t clock_id;
+};
+
+struct scmi_clock_duty_cycle_get_p2a {
+	int32_t status;
+	uint32_t num;
+	uint32_t den;
+};
+
 /*
  * Clock Rate Set
  */
diff --git a/core/include/drivers/scmi-msg.h b/core/include/drivers/scmi-msg.h
index c276ea853..a921c0f7d 100644
--- a/core/include/drivers/scmi-msg.h
+++ b/core/include/drivers/scmi-msg.h
@@ -286,6 +286,11 @@ int32_t plat_scmi_clock_get_state(unsigned int channel_id,
 int32_t plat_scmi_clock_set_state(unsigned int channel_id, unsigned int scmi_id,
 				  bool enable_not_disable);
 
+struct clk_duty;
+
+int32_t plat_scmi_clock_get_duty_cycle(unsigned int channel_id, unsigned int scmi_id,
+				       struct clk_duty *duty);
+
 /* Handlers for SCMI Reset Domain protocol services */
 
 /*
-- 
2.47.2

